app/models/grda_warehouse/tasks/client_cleanup.rb

```diff
@@ -219,13 +221,77 @@ def choose_best_gender dest_attr, source_clients
       # Get the best Gender (has 0..4, newest breaks the tie)
       known_values = [0, 1, 2, 3, 4]
       known_value_gender_clients = source_clients.select{|sc| known_values.include?(sc[:Gender])}
-      if !known_values.include?(dest_attr[:Gender]) or known_value_gender_clients.any?
+      if !known_values.include?(dest_attr[:Gender]) || known_value_gender_clients.any?
         known_value_gender_clients = source_clients if known_value_gender_clients.none? #if none have known values we consider them all in the sort test
         dest_attr[:Gender] = known_value_gender_clients.sort{|a, b| a[:DateUpdated] <=> b[:DateUpdated]}.last[:Gender]
       end
       dest_attr
     end
 
+    def choose_best_race dest_attr, source_clients
+      # Most recent 0 or 1 if no 0 or 1 use the most recent value
+      # Valid responses for race categories are [0, 1, 99]
+      # Valid responses for RaceNone are [8, 9, 99] -- should be null if all other fields are 0 or 99
+      known_values = [0, 1]
+      # Sort in reverse chronological order (newest first)
+      sorted_source_clients = source_clients.sort_by.sort{|a, b| b[:DateUpdated] <=> a[:DateUpdated]}
+
+      race_columns.each do |col|
+        sorted_source_clients.each do |source_client|
+          value = source_client[col]
+          current_value = dest_attr[col]
+          # if we have a 0 or 1 use it
+          # otherwise only replace if the current value isn't a 0 or 1
+          dest_attr[col] = if known_values.include?(value)
+            value
+          elsif !known_values.include?(current_value)
+            value
+          else
+            current_value
+          end
+
+          # Since these are sorted in reverse chronological order, if we hit a 1 or 0, we'll consider that
+          # the destination client response
+          break if known_values.include?(value)
+
+        end
+      end
+      # if we still don't have any responses, use the most-recent RaceNone response
+      dest_attr[:RaceNone] = sorted_source_clients.first[:RaceNone] if dest_attr.values_at(*race_columns).any?(1)
+      dest_attr
+    end
+
+    private def race_columns
+      @race_columns ||= GrdaWarehouse::Hud::Client.race_fields.map(&:to_sym) - [:RaceNone]
+    end
+
+    def choose_best_ethnicity dest_attr, source_clients
+      # Most recent 0 or 1 if no 0 or 1 use the most recent value
+      known_values = [0, 1]
+      # Sort in reverse chronological order (newest first)
+      sorted_source_clients = source_clients.sort_by.sort{|a, b| b[:DateUpdated] <=> a[:DateUpdated]}
+      col = :Ethnicity
```

Comment:
If the column is set as a variable anyway, can we DRY anything else up by extracting out `choose_best_column_value` or such like?

Reply:
Yes, we could, though the logic is only the same for a few of them, and historically people have wanted different logic for each type.  I'm inclined to leave it for now, but this is good to note for later.